README.md
<center>

# TemperatureMonitor

</center>

# Releases

## `v0.2.1-stable`

* Added color coding of temperature levels

## `v0.1.2-stable`

* First release on 07/23/2022

<hr>

# Program Description
Reads system temperature from `/sys/class/thermal/thermal_zone0/temp` and outputs to human readable format in Celsius (&deg;C), Fahrenheit (&deg;F), and Kelvin (K).  This is a terminal based application is intended for exclusive use on Linux machines and will not work on Windows or Mac.

Please report issues.

This software is FOSS (Free and Open Source Software).  Therefore there is NO COST to download and use this software.  If you would like to support my work you can [Buy Me a Coffee](https://www.buymeacoffee.com/kwilkins).

<hr>

# Screenshots

![Preview](https://raw.githubusercontent.com/kawilkins/TemperatureMonitor/0.2.1/screenshots/tempmon-preview01.png)

![Preview](https://raw.githubusercontent.com/kawilkins/TemperatureMonitor/0.2.1/screenshots/tempmon-preview02.png)

![Preview](https://raw.githubusercontent.com/kawilkins/TemperatureMonitor/0.2.1/screenshots/tempmon-preview03.png)

# Quick setup

**Download binary from github**
```
wget https://github.com/kawilkins/TemperatureMonitor/releases/download/v0.2.1-stable/tempmon
```

**Make it executable**
```
chmod +x tempmon
```
**Move it to ```/usr/bin```**
```
sudo mv tempmon /usr/bin
```
**Test it out**
```
 tempmon
```
<hr>

# Compiling
It is easy to Compile but you need to have g++

**Clone the repo with git**
```
git clone https://github.com/kawilkins/TemperatureMonitor.git
```
**Change directory and then compile it**
```
cd TemperatureMonitor/src
g++ -o tempmon tempmonCPU.cpp
```
Additionally you may want to move the complied binary to ```/usr/bin```.


CHANGELOG.md
- Updated syntax to standard. Remove 'using namespace std;'. Remove unnecessary white space around '{}'.
- Removed Test directory. Created TEST branch.
- Updated change log.
- Updated README with current release information.
- Merge branch '0.2.1'
- Updated README with screen shots showing results from testing.
- Verified test data to ensure proper color code thresholds.
- Changed script to simply check for a test file and compile instead of copying and compiling the same base file all the time.
- Add screen shots.
- Changed order of commands in script.
- Add testing application.
- Change name of testing script.
- Renamed tempmonCPU.cpp -> tempmon.cpp.
- Add test script to download and compile cpp file.
- Add center alignment to program title. Update program description to include temperature units output. Add section for release descriptions. Add screenshot preview.
- Add screen shot of output.
- Updated LICENSE with correct wording.
- Updated LICENSE with correct wording.
- Updated change log with recent commit.
- Updated change log with recent commit.
- Remove excessive and unnecessary comments. Add definition for double line spacing. Add definition for color formatting. Remove unnecessary variable that converted string to double. Simplified Celsius conversion. Formatted spacing and alignment of output. Add evaluation of temperature data and corresponding text color. Add status output. Add message to user.
- Improved formatting of output using 'setw()' and 'right' manipulators from iomanip.
- Add "fixed" precision to output.
- Update README.md
- Update README.md
- Updated change log.
- Update CHANGELOG.md
- Update CHANGELOG.md
- Adding changelog.
- Merge branch 'Darin755/master' Merging and approving proposed change to README.md with edit to FOSS definition.
- Remove "FLOSS". Add FOSS and definition.
- Delete README.md.backup
- Update README.md
- Update README.md
- fixed formating
- improved readme
- Add release information.
- Corrected version date.
- Updated project description. Add guide to installation.
- Corrected version date.
- Moved temperature conversion variables to while loop. Remove 'tempFahrenheit = tempCelsius * (9/5) + 32'. Add 'tempFahrenheit = (tempCelsius * 1.8) + 32'
- Created GNU General Public License version 3 (GPLv3). Removed template information. Added program name. Added program description.
- Added comment before thermal.open() function. Compiled and test run. Program compiled but failed to run.
- Added comment before thermal.open() function. Attempted debug and received error:
- Removed unnecessary comments. Cleaned up code spacing.
- Corrected version number in description.
- Determined that separate function was overkill. Merged operation into main(). Removed Thermal() function. Added variables to output system temp data to:  * Degrees Celsius  * Degrees Fahrenheit  * Kelvin Added 'ios::in' to thermal.open() correcting syntax error.
- Added statement to convert string to double. Added statement to divide temperature output and store result in variable to be output to the screen in degrees C. Added iomanip and string libraries.
- Corrected syntax errors.
- Corrected variable syntax error.
- Created function for fetching information from /sys/class/thermal/thermal_zone0/temp.
- Updated application name.
- Added new file tempmonCPU.cpp.
- Delete LICENSE
- Initial commit


Test/tempmonTEST.sh
#!/bin/bash

# Set current version and file for testing
VERSION = 0.2.1
FILE = tempmonCPU.cpp

# Download and compile
wget https://raw.githubusercontent.com/kawilkins/TemperatureMonitor/$(VERSION)/src/tempmonCPU.cpp
g++ -o tempmonTEST $(FILE)

Test/tempmonTEST.cpp
/*==============================================================
Program:  CPU Temperature Monitor
Version:  0.2.1
Version Date:  09/06/2022
Author:  Kevin Wilkins
Date:  02/05/2022
Parameters:
This application will read out CPU temperature data from 
'/sys/class/thermal/thermal_zone0/temp' file on Linux and will
convert output to human readable format in degrees Celsius, 
degrees Fahrenheit, and Kelvin.
==============================================================*/

// Preprocessor directives and namespace usage
#include <iostream>
#include <iomanip>
#include <fstream>
#include <string>

using namespace std;

// Special definitions
#define DOUBLELINE "\n\n"
#define RESTORE "\033[0m" // Return to default terminal color
#define GREEN "\033[1m\033[32m" // Bold green
#define YELLOW "\033[1m\033[33m" // Bold yellow
#define RED "\033[1m\033[31m" // Bold red

// Main function
int main()
{
    ifstream thermal;
    string sysTemp = "";
    
    // Read in /sys/.../temp file
    thermal.open("/sys/class/thermal/thermal_zone0/temp", ios::in);

    // Check if file is open and output in human readable format
    if (thermal.is_open())
    {
        while (getline (thermal, sysTemp))
        {
            // Convert output
            // double tempCelsius = stod(sysTemp) / 1000;
            
            double tempCelsius = 87; //Test value

            double tempFahrenheit = (tempCelsius * 1.8) + 32;
            double tempKelvin = tempCelsius + 273.15;
            string colorCode = "";
            string status = "";
            string message = "";
            
            // Evaluate Temperature Safety Level
            if(tempCelsius < 60)
            {
                colorCode = GREEN;
                status = "GOOD";
                message = "CPU temperature is very good.";
            }
            else if(tempCelsius >= 60 && tempCelsius < 75)
            {
                colorCode = YELLOW;
                status = "CAUTION";
                message = "CPU temperature is somewhat elevated.";
            }
            else if(tempCelsius >= 75)
            {
                colorCode = RED;
                status = "WARNING";
                message = "CPU temperature is high and needs attention.";
            }
            
            // Output to terminal as human readable
            cout << fixed << setprecision(3);
            cout << "System Temperature\n";
            cout << "Status:" << colorCode << status << RESTORE;
            cout << DOUBLELINE;
            cout << colorCode;
            cout << setw(10) << right << tempCelsius << " \u00B0C\n"
                 << setw(10) << right << tempFahrenheit << " \u00B0F\n"
                 << setw(10) << right << tempKelvin << "  K" << DOUBLELINE;
            cout << RESTORE;
            cout << message << DOUBLELINE;
        }
        thermal.close();
    }
    else
    {
        cout << "Failure to open '/sys/class/thermal/thermal_zone0/temp'\n"
             << "Please try again . . ."
             << DOUBLELINE;
    }

    // Exit program without errors
    return 0;
}

Test/TestCompileRun.sh
#!/bin/bash

TESTFILE=tempmonTEST.cpp
SOURCE=../src/tempmon.cpp

if [ -f $TESTFILE ]; then
    g++ $TESTFILE --verbose -o tempmonTEST
    printf "\n\n\xE2\x9C\x94 \e[1;32mExecutable 'tempmonTEST' created . . . \n\n\e[0m"
else
    cp $SOURCE $TESTFILE
    g++ $TESTFILE --verbose -o tempmonTEST
    printf "\n\n\xE2\x9C\x94 \e[1;32mtempmonTEST.cpp created . . . \n\e[0m"
    printf "\xE2\x9C\x94 \e[1;32mExecutable 'tempmonTEST' created . . . \n\n\e[0m"
fi


src/tempmonCPU.cpp
/*==============================================================
Program:  CPU Temperature Monitor
Version:  0.2.1
Version Date:  09/15/2022
Author:  Kevin Wilkins
Date:  02/05/2022
Parameters:
This application will read out CPU temperature data from 
'/sys/class/thermal/thermal_zone0/temp' file on Linux and will
convert output to human readable format in degrees Celsius, 
degrees Fahrenheit, and Kelvin.
==============================================================*/

// Preprocessor directives and namespace usage
#include <iostream>
#include <iomanip>
#include <fstream>
#include <string>

using namespace std;

// Main function
int main()
{
    ifstream thermal;
    string sysTemp = ""; // Will store the output of thermal
    
    // Read in /sys/.../temp file
    // Point read pointer to beginning of file
    thermal.open ("/sys/class/thermal/thermal_zone0/temp", ios::in);

    // Check if file is open and output in human readable format
    if (thermal.is_open())
    {
        while (getline (thermal, sysTemp))
        {
            // Convert output
            double tempLong = stod(sysTemp); // Convert string to double
            double tempCelsius = tempLong / 1000; // Convert output to human readable
            double tempFahrenheit = (tempCelsius * 1.8) + 32; // Convert to Fahrenheit
            double tempKelvin = tempCelsius + 273.15; // Convert to Kelvin
            
            // Output to human readable format
            cout << fixed << setprecision(3); // Use 3 significant digits
            cout << "System temperature:  \n"
                 << setw(10) << right << tempCelsius << " \u00B0C\n"
                 << setw(10) << right << tempFahrenheit << " \u00B0F\n"
                 << setw(10) << right << tempKelvin << "  K\n";
        }
        thermal.close();
    }
    else // File does not open
    {
        cout << "Failure to open '/sys/class/thermal/thermal_zone0/temp'\n"
             << "Please try again . . .\n\n";
    }

    // Exit program without errors
    return 0;
}


src/tempmon.cpp
/*==============================================================
Program:  CPU Temperature Monitor
Version:  0.2.2
Version Date:  01/08/2023
Author:  Kevin Wilkins
Date:  02/05/2022
Parameters:
This application will read out CPU temperature data from
'/sys/class/thermal/thermal_zone0/temp' file on Linux and will
convert output to human readable format in degrees Celsius, 
degrees Fahrenheit, and Kelvin.
==============================================================*/

// Preprocessor directives and namespace usage
#include <iostream>
#include <iomanip>
#include <fstream>
#include <string>

// Special definitions
#define DOUBLELINE "\n\n"
#define RESTORE "\033[0m" // Return to default terminal color
#define GREEN "\033[1m\033[32m" // Bold green
#define YELLOW "\033[1m\033[33m" // Bold yellow
#define RED "\033[1m\033[31m" // Bold red

// Main function
int main(){
    std::ifstream thermal;
    std::string sysTemp = "";
    
    // Read in /sys/.../temp file
    thermal.open("/sys/class/thermal/thermal_zone0/temp", std::fstream::in);

    // Check if file is open and output in human readable format
    if (thermal.is_open()){
        while (getline (thermal, sysTemp)){
            // Convert output
            double tempCelsius = stod(sysTemp) / 1000;
            double tempFahrenheit = (tempCelsius * 1.8) + 32;
            double tempKelvin = tempCelsius + 273.15;
            std::string colorCode = "";
            std::string status = "";
            std::string message = "";
            
            // Evaluate Temperature Safety Level
            if(tempCelsius < 60){
                colorCode = GREEN;
                status = "GOOD";
                message = "CPU temperature is very good.";
            }else if(tempCelsius >= 60 && tempCelsius < 75){
                colorCode = YELLOW;
                status = "CAUTION";
                message = "CPU temperature is somewhat elevated.";
            }else if(tempCelsius >= 75){
                colorCode = RED;
                status = "WARNING";
                message = "CPU temperature is high and needs attention.";
            }
            
            // Output to terminal as human readable
            std::cout << std::fixed << std::setprecision(3);
            std::cout << "System Temperature\n";
            std::cout << "Status:" << colorCode << status << RESTORE;
            std::cout << DOUBLELINE;
            std::cout << colorCode;
            std::cout << std::setw(10) << std::right << tempCelsius << " \u00B0C\n";
            std::cout << std::setw(10) << std::right << tempFahrenheit << " \u00B0F\n";
            std::cout << std::setw(10) << std::right << tempKelvin << "  K";
            std::cout << DOUBLELINE << RESTORE << message << DOUBLELINE;
        }
        thermal.close();
    }else{
        std::cout << "Failure to open '/sys/class/thermal/thermal_zone0/temp'\n";
        std::cout << "Please try again . . ." << DOUBLELINE;
    }

    // Exit program without errors
    return 0;
}